merlion/models/forecast/deep_base.py

```diff
@@ -0,0 +1,261 @@
+                self.optimizer.zero_grad()
+
+                loss, _, _ = self._get_batch_model_loss_and_outputs(self._convert_batch_to_tensors(batch))
+                train_loss.append(loss.item())
+
+                loss.backward()
+                if config.clip_gradient is not None:
+                    torch.nn.utils.clip_grad_norm(self.model.parameters(), config.clip_gradient)
+
+                self.optimizer.step()
+
+            train_loss = np.average(train_loss)
+
+            # set validation flag
+            total_dataset.validation = True
+            _, val_loss = self._get_np_loss_and_prediction(total_dataset)
+            total_dataset.validation = False
+
+            if bar is not None:
+                bar.print(
+                    epoch + 1, prefix="", suffix=f"Train Loss: {train_loss: .4f}, Validation Loss: {val_loss: .4f}"
+                )
+
+            if config.early_stop_patience:
+                early_stopping(val_loss, self.deep_model)
+                if early_stopping.early_stop:
+                    logger.info(f"Early stopping with {config.early_stop_patience} patience")
+                    break
+
+        if config.early_stop_patience:
+            early_stopping.load_best_model(self.deep_model)
+            logger.info(f"Load the best model with validation loss: {early_stopping.val_loss_min: .4f}")
+
+        logger.info("End of the training loop")
+
+        pred, _ = self._get_np_loss_and_prediction(total_dataset)
+
+        # since the model predicts multiple steps, we concatenate all the first steps together
+        return pd.DataFrame(pred[:, 0], columns=train_data.columns), None
+
+    def _get_batch_model_loss_and_outputs(self, batch):
+        """
+        For loss calculation and output prediction
+
+        :param batch: a batch contains `(past, past_timestamp, future, future_timestamp)` used for calculating loss and model outputs
+
+        :return: calculated loss, deep model outputs and targeted ground truth future
+        """
+        config = self.config
+        past, past_timestamp, future, future_timestamp = batch
+
+        model_output = self.deep_model(past, past_timestamp, future_timestamp)
+
+        if future is None:
+            return None, model_output, None
+
+        if config.target_seq_index is None and self.support_multivariate_output:
+            target_future = future
+        else:
+            # choose specific target_seq_index for regression
+            target_idx = config.target_seq_index
+            target_future = future[:, :, target_idx : target_idx + 1]
+
+        loss = self.loss_fn(model_output, target_future)
+        return loss, model_output, target_future
+
+    @property
+    def require_even_sampling(self) -> bool:
+        return False
+
+    def _forecast(
+        self, time_stamps: List[int], time_series_prev: pd.DataFrame, return_prev=False
+    ) -> Tuple[pd.DataFrame, Optional[pd.DataFrame]]:
+
```

Comment:
You need to handle the case where `time_series_prev` is None. The standard way to do this would be to set `time_series_prev = self.transform(self.train_data).to_pd()`.